src/PowerShellEditorServices.Hosting/HostLogger.cs

```diff
@@ -0,0 +1,332 @@
+        {
+            if (observer == null)
+            {
+                throw new ArgumentNullException(nameof(observer));
+            }
+
+            return Subscribe(new LogObserver(observer));
+        }
+
+        /// <summary>
+        /// Log a message to log sinks.
+        /// </summary>
+        /// <param name="logLevel">The log severity level of message to log.</param>
+        /// <param name="message">The message to log.</param>
+        public void Log(PsesLogLevel logLevel, string message)
+        {
+            // Do nothing if the severity is lower than the minimum
+            if (logLevel < _minimumLogLevel)
+            {
+                return;
+            }
+
+            // Remember this for later subscriptions
+            _logMessages.Enqueue((logLevel, message));
+
+            // Send this log to all observers
+            foreach (IObserver<(PsesLogLevel logLevel, string message)> observer in _observers.Keys)
+            {
+                observer.OnNext((logLevel, message));
+            }
+        }
+
+        /// <summary>
+        /// Convenience method for logging exceptions.
+        /// </summary>
+        /// <param name="message">The human-directed message to accompany the exception.</param>
+        /// <param name="exception">The actual execption to log.</param>
+        /// <param name="callerName">The name of the calling method.</param>
+        /// <param name="callerSourceFile">The name of the file where this is logged.</param>
+        /// <param name="callerLineNumber">The line in the file where this is logged.</param>
+        public void LogException(
+            string message,
+            Exception exception,
+            [CallerMemberName] string callerName = null,
+            [CallerFilePath] string callerSourceFile = null,
+            [CallerLineNumber] int callerLineNumber = -1)
+        {
+            Log(PsesLogLevel.Error, $"{message}. Exception logged in {callerSourceFile} on line {callerLineNumber} in {callerName}:\n{exception}");
+        }
+
+    }
+
+    /// <summary>
+    /// A log sink to direct log messages back to the PowerShell host.
+    /// </summary>
+    /// <remarks>
+    /// Note that calling this through the cmdlet causes an error,
+    /// so instead we log directly to the host.
+    /// Since it's likely that the process will end when PSES shuts down,
+    /// there's no good reason to need objects rather than writing directly to the host.
+    /// </remarks>
+    internal class PSHostLogger : IObserver<(PsesLogLevel logLevel, string message)>
+    {
+        private readonly PSHostUserInterface _ui;
+
+        /// <summary>
+        /// Create a new PowerShell host logger.
+        /// </summary>
+        /// <param name="ui">The PowerShell host user interface object to log output to.</param>
+        public PSHostLogger(PSHostUserInterface ui)
+        {
+            _ui = ui;
+        }
+
+        public void OnCompleted()
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation.](https://app.codacy.com/app/TylerLeonhardt/PowerShellEditorServices/pullRequest?prid=4579423)